flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2159,138 +2159,265 @@ struct CoordinateOpConversion
     }
-    if (!hasSubdimension)
-      columnIsDeferred = true;
 
-    if (!validCoordinate(cpnTy, operands.drop_front(1)))
-      return mlir::emitError(loc, "coordinate has incorrect dimension");
+    rewriter.replaceOpWithNewOp<mlir::LLVM::BitcastOp>(coor, ty, resultAddr);
+    return success();
+  }
+
+  mlir::LogicalResult
+  doRewriteRefOrPtr(fir::CoordinateOp coor, mlir::Type ty,
+                    mlir::ValueRange operands, mlir::Location loc,
+                    mlir::ConversionPatternRewriter &rewriter) const {
+    mlir::Type baseObjectTy = coor.getBaseType();
+
+    mlir::Type currentObjTy = fir::dyn_cast_ptrOrBoxEleTy(baseObjectTy);
+    bool hasSubdimension = hasSubDimensions(currentObjTy);
+    bool columnIsDeferred = !hasSubdimension;
+
+    if (!supportedCoordinate(currentObjTy, operands.drop_front(1)))
+      TODO(loc, "unsupported combination of coordinate operands");
 
-    // if arrays has known shape
     const bool hasKnownShape =
-        arraysHaveKnownShape(cpnTy, operands.drop_front(1));
+        arraysHaveKnownShape(currentObjTy, operands.drop_front(1));
 
     // If only the column is `?`, then we can simply place the column value in
     // the 0-th GEP position.
-    if (auto arrTy = cpnTy.dyn_cast<fir::SequenceType>()) {
+    if (auto arrTy = currentObjTy.dyn_cast<fir::SequenceType>()) {
       if (!hasKnownShape) {
-        const auto sz = arrTy.getDimension();
+        const unsigned sz = arrTy.getDimension();
         if (arraysHaveKnownShape(arrTy.getEleTy(),
                                  operands.drop_front(1 + sz))) {
-          auto shape = arrTy.getShape();
+          llvm::ArrayRef<int64_t> shape = arrTy.getShape();
           bool allConst = true;
-          for (std::remove_const_t<decltype(sz)> i = 0; i < sz - 1; ++i)
+          for (unsigned i = 0; i < sz - 1; ++i) {
             if (shape[i] < 0) {
               allConst = false;
               break;
             }
+          }
           if (allConst)
             columnIsDeferred = true;
         }
       }
     }
 
-    if (fir::hasDynamicSize(fir::unwrapSequenceType(cpnTy)))
-      return mlir::emitError(
+    if (fir::hasDynamicSize(fir::unwrapSequenceType(currentObjTy))) {
+      mlir::emitError(
           loc, "fir.coordinate_of with a dynamic element size is unsupported");
+      return failure();
+    }
 
     if (hasKnownShape || columnIsDeferred) {
       SmallVector<mlir::Value> offs;
-      if (hasKnownShape && hasSubdimension)
+      if (hasKnownShape && hasSubdimension) {
+        mlir::LLVM::ConstantOp c0 =
+            genConstantIndex(loc, lowerTy().indexType(), rewriter, 0);
         offs.push_back(c0);
-      const auto sz = operands.size();
+      }
+      const std::size_t sz = operands.size();
       Optional<int> dims;
       SmallVector<mlir::Value> arrIdx;
-      for (std::remove_const_t<decltype(sz)> i = 1; i < sz; ++i) {
-        auto nxtOpnd = operands[i];
+      for (std::size_t i = 1; i < sz; ++i) {
```

Comment:
And here. Can `sz` be moved into the loop? (It wasn't before because `i` was declared to be whatever type `sz` was and without a `const`.